### DISCUSSION TITLE
How to set the relative coordinate from the other components

### DISCUSSION THREAD

**Discussion by inse0918** [byuflowlab/FLOWUnsteady#160] [Show and tell]
Hi all.

Is there any way to set the relative coordinate from the components ?

For example, rotor coordinate can be set like below figure.

![image](https://github.com/byuflowlab/FLOWUnsteady/assets/93040154/1c327850-4e7a-44f7-9814-e611ddcd9e13)

I want to tilt the rotor_1 and rotor_2 relative to local_1 and local_2.

But all the rotor tilt relative to global coordinate like below figure.

![image](https://github.com/byuflowlab/FLOWUnsteady/assets/93040154/1de42e81-a76c-40a7-bf6a-bc9cfcd96f90)

Thank you.
Best regards,
Inseo.

**Comment by inse0918** [comment]
![image](https://github.com/byuflowlab/FLOWUnsteady/assets/93040154/9fe1c8bf-0242-426d-b483-2bc8ccfdc255)

[Generating components](https://flow.byu.edu/FLOWUnsteady/api/flowunsteady-vehicle-components/#FLOWVLM-Systems) in FLOWunsteady doc.

**Reply by EdoAlvarezR** [reply]
Hi Inseo!

Using the first rotor as an example, I would first move the rotor away from its local origin
```julia
import FLOWUnsteady as uns
import FLOWUnsteady: vlm
.
.
.
O = rotor_1 - local_1
Oaxis = uns.I
vlm.setcoordsystem(rotor, O, Oaxis; user=true)
```
then I would add the rotor to a WingSystem that will tilt and move to its desired origin:
```julia
tiltsystem = vlm.WingSystem()
vlm.addwing(tiltsystem, "Rotor1", rotor)

O = local_1
vlm.setcoordsystem(tiltsystem, O, Oaxis)
```
finally, I would add that to the overall vehicle as a tilting system with the desired tilting motion.

Hopefully this helps, but let me know if you get stuck.

**Reply by inse0918** [reply]
Hi Edo.

Thank you for replying kindly answer although the discussion was closed.

I used below line for applying the relative coordinate system.

  ```
  O = [0.0, 0.0, 0.0]                                 # New position
  Oaxis = uns.gt.rotation_matrix2(0.0, 0.0, 180)     # New orientation
  vlm.setcoordsystem(rotor, O, Oaxis)
  ```
First, It means that the `rotor` in a coordinate defined as `(0, 0, 0)` was created first.

  ```
  heavingsystem1 = vlm.WingSystem(Any[], String[], [1.0, 0.0, 0.0],
                                  [1.0 0.0 0.0; 0.0 1.0 0.0; 0.0 0.0 1.0],
                                  [1.0 0.0 0.0; 0.0 1.0 0.0; 0.0 0.0 1.0],
                                  nothing, Dict{String, Any}())  
  vlm.addwing(heavingsystem1, "rotor", rotor)
  ```
Next, the `heavingsystem1` with no components was defined as `(1.0, 0.0, 0.0)`, so the fixed coordinate system was created.

  `tilting_systems = (heavingsystem1, );        # Tilting systems `
And next, the `rotor` was put in on the `heavingsystem1`, so the relative coordinate of the `rotor` position is distance with +1.0 in `x-axis`.

And also it means that, the `heavingsystem1` act like a pivot point.

  ```
  system = vlm.WingSystem()                   # System of all FLOWVLM objects
  vlm.addwing(system, "Rotor", heavingsystem1)
  ```
Finally, all the relative coordinate system was put in the `system = vlm.WingSystem()` for arrange the components based global coordinate.

Above the contents is what I understand the `FLOWVLM.addwing` and the `FLOWVLM.setcoordsystem`.
 
And the simulation worked well as I intended.

Thank you.
Best regards,
Inseo.

**Reply by EdoAlvarezR** [reply]
Awesome, thanks for sharing your solution!

**Reply by inse0918** [reply]
Thank you Edo.

I have one more question about the calculating the forces.

If the incidence angle(tilt angle in my case) of the propeller is changing, the force from the propeller will be change too.

The post-processed force and torque can be defined in the `FLOWUnsteady_monitors.jl`, also it was extended from `CCBlade.jl` initially.

**1. Is there any way to export the Fx, Fy, Fz, Tx, Ty, Tz of the propeller ?**

For now, the force and torque component of the propeller can be checked in *.csv file using CT and CQ.

But this data have a limitation that can not considered 3-axis analysis by tilt angle, because this data is relative with the rotor plane axis. 

When I checked the `CCBlade.jl' there is no Y-axis components about the propeller force and torque.

When I want to export the X,Y,Z axis forces and torques, the `DistributedLoad` was used which can be extract from the `*.vtk` files using the paraview analysis. But this way is very complex, because this way is not using the originally calculated force of the propeller but using the post-processed data.

For example, if the Fx, Fy, Fz, Tx, Ty, Tz can be extracted, the below analysis will be enable
_Stokkermans, T.C.A.; Veldhuis, L.L.M., Propeller Performance at Large Angle of Attack Applicable to Compound Helicopters_

![image](https://github.com/byuflowlab/FLOWUnsteady/assets/93040154/1372cd0f-b8fa-4fde-a2a7-3c46f17a6706)

I hope there is a way to extract 3-axis data.

Thank you.
Best regards,
Inseo.

**Reply by ilm71** [reply]
Dear All, 

I would be really interested in such a function too where I can export in a clear way the Fx, Fy, Fz, Tx, Ty, Tz of the propeller. 
Thanks for your reply!

Best

**Reply by inse0918** [reply]
Hi, now I'm working on a validation about the previous force/torque [postprocess code on MATLAB](https://github.com/byuflowlab/FLOWUnsteady/issues/109#issuecomment-2051353668).

First, the problem of the previous code is too much time consuming in extracting the `DistributedLoad` using Paraview macro.

So, I'm going to implement a script into the `FLOWVLM_rotor` for extracting the `DistributedLoad` directly when the analysis is running.

As I found the data of the `DistributedLoad`, the force/torque can be found at the 
 ```
Fs = [ _ccblade2global( get_blade(self, blade_i), ccb_F, self.CW;
                          translate=false) for ccb_F in ccb_Fs ]
```

Thank you.
Best regards,
Inseo.

**Reply by ilm71** [reply]
Hi Inseo, 
thanks for the quick reply. If I understand you correctly, you have implemented a separate routine in the FLOWVLM_rotor. Might you share the whole final scripts here to get a full understanding what has been changed? I would be very glad to test your scripts for my cases. Thanks for your answer and your efforts!!

Best

**Reply by inse0918** [reply]
Hi.

Actually, I did not add the isolated subroutine into the FLOWVLM_rotor. I just extract the data that I want.

So, the problem of the previous method using data by *.vtk extraction with Paraview was solved(too slow and complex for extracting the distributed load, else.).

But, the method using Matlab still available and I'm in a plan about making a isolated subroutine or function into the FLOWVLM_rotor by Julia for user soon.

The one of the worries is if there is a effect of the extracting data on calculating the whole simulation time.

Anyway, I'll compare the tilt 90 deg result soon(maybe today).

I'll upload the modified FLOWVLM_rotor on the other [discussion posting](https://github.com/byuflowlab/FLOWUnsteady/discussions/163).

May I ask you, can you check if calculating the Fx,y,z and Mx, y, z in my Matlab code was scripted in a correct step(formulation) ?

Thank you.
Best regards,
Inseo.

**Reply by ilm71** [reply]
Hi @inse0918, any updates regarding the forces of the tilted rotor and in addition the forces on each of the distributed rotors? Have you adapted therefore your julia-file? Thanks!
